File: August2022/Java/RomanToInteger.java

// @saorav21994
// TC : O(lengthOfString)
// SC : O(1)

class Solution {
    public int romanToInt(String s) {
        int res = 0;
        int l = s.length();
        char ch = s.charAt(l-1);
        if (ch == 'I') res += 1;
        else if (ch == 'V') res += 5;
        else if (ch == 'X') res += 10;
        else if (ch == 'L') res += 50;
        else if (ch == 'C') res += 100;
        else if (ch == 'D') res += 500;
        else if (ch == 'M') res += 1000;
        for (int i = l-2; i >= 0; i--) {
            ch = s.charAt(i);
            if (ch == 'M') res += 1000;
            else if (ch == 'D') {
                if (s.charAt(i+1) == 'M') res -= 500;
                else res += 500;
            }
            else if (ch == 'C') {
                if (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M') res -= 100;
                else res += 100;
            }
            else if (ch == 'L') {
                if (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M' || s.charAt(i+1) == 'C') res -= 50;
                else res += 50;
            }
            else if (ch == 'X') {
                if (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M' || s.charAt(i+1) == 'C' || s.charAt(i+1) == 'L') res -= 10;
                else res += 10;
            }
            else if (ch == 'V') {
                if (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M' || s.charAt(i+1) == 'C' || s.charAt(i+1) == 'L' || s.charAt(i+1) == 'X') res -= 5;
                else res += 5;
            }
            else if (ch == 'I') {
                if (s.charAt(i+1) == 'D' || s.charAt(i+1) == 'M' || s.charAt(i+1) == 'C' || s.charAt(i+1) == 'L' || s.charAt(i+1) == 'X' || s.charAt(i+1) == 'V') res -= 1;
                else res += 1;
            }
        }
        return res;
    }
}

File: August2022/C++/RomanToInteger.cpp
class Solution {
public:
    // TC: O(n)
    // SC: O(1)
    int romanToInt(string s) {
        unordered_map<char, int> map;
        map['I'] = 1;
        map['V'] = 5;
        map['X'] = 10;
        map['L'] = 50;
        map['C'] = 100;
        map['D'] = 500;
        map['M'] = 1000;
        
        int num = 0;
        
        int currNum;
        char c, nextChar;        
        for (int i = s.size() - 1; i >= 0; i--) {
            c = s[i];
            currNum = map[c];
            
            if (i != s.size() - 1) {
                nextChar = s[i + 1];
                
                if (c == 'I' && (nextChar == 'V' || nextChar == 'X')) {
                    num -= currNum;
                } else if (c == 'X' && (nextChar == 'L' || nextChar == 'C')) {
                    num -= currNum;
                } else if (c == 'C' && (nextChar == 'D' || nextChar == 'M')) {
                    num -= currNum;
                } else {
                    num += currNum;
                }
            } else {
                num += currNum;
            }
        }
        
        return num;
    }
};
